auth.py

```diff
@@ -0,0 +1,304 @@
+        
+        # Define headers to request the token
+        headers = {"Content-Type": "application/soap+xml; charset=utf-8"}
+
+        # Insert variables into SAML request
+        password = self.password or getpass("Enter your password: ")
+        saml = saml.format(username=escape(self.username),
+                           password=escape(password),
+                           auth_url=self.auth_url,
+                           message_id=message_id,
+                           created=createdStr,
+                           expires=expiresStr)
+        
+        # Request security token from Microsoft Online
+        print("Requesting security token...\r", end="")
+        try:
+            response = requests.post(self.auth_url, data=saml, headers=headers)
+        except requests.exceptions.ConnectionError:
+            print("Could not connect to", self.auth_url)
+            return
+        # Parse and extract token from returned XML
+        try:
+            root = et.fromstring(response.text)
+        except et.ParseError:
+            print("Token request failed. The server did not send a valid response\r", end="")
+            return
+        try:
+            samlAssertion = (re.search('<saml:Assertion.*\/saml:Assertion>', response.text)).group(0)
+        except AttributeError:
+            print("Token requested, but response did not include the STS SAML Assertion.\r", end="")
+            return
+        
+        print("SAML Assertion received.")
+
+        # Get the BinarySecurityToken
+
+        # Define MS Online authentication REST endpoint
+        MSO_AUTH_URL = "https://login.microsoftonline.com/rst2.srf"
+
+        with open(os.path.join(os.path.dirname(__file__), 
+                "saml-templates/sp-adfs-stsBinaryToken.xml"), "r") as file:
+            saml = file.read()
+
+        # Format saml to receive BinarySecurityToken
+        saml = saml.format(customSTSAssertion=samlAssertion,
+                           msoEndpoint="sharepoint.com")
+        response = requests.post(url=MSO_AUTH_URL, data=saml, headers=headers)
+        # Parse and extract token from returned XML
+        try:
+            root = et.fromstring(response.text)
+        except et.ParseError:
+            print("BinarySecurityToken request failed. The server did not send a valid response\r", end="")
+            return
+        try:
+            binarySecurityToken = (re.search('BinarySecurityToken Id.*>([^<]+)', response.text)).group(1)
+        except AttributeError:
+            print("BinarySecurityToken requested, but response did not include the STS SAML Assertion.\r", end="")
+            return
+        
+        print("BinarySecurityToken received.")
+        self.token = binarySecurityToken
+
+        return True
+    
+    def _get_cookie(self):
+        # regex will get the 'tenant url' from the site provided for login
+        SPO_IDCRL_URL = "https://" + re.search('(.+\.com)/', self.site).group(1) + "/_vti_bin/idcrl.svc/"
```

Comment:
Same here

Reply:
✔️ 

this is the same regex for `self.tenantBaseUrl` anyway, so I've just changed the regex here to use that variable instead.